napari/_vispy/vispy_surface_layer.py

```diff
@@ -51,8 +45,8 @@ def _on_slice_data_change(self, event=None):
         self._on_scale_change()
         self._on_translate_change()
 
-    def _on_colormap_change(self, event=None):
-        cmap = self.layer.colormap[1]
+    def _on_colormap_change(self, colormap):
```

Comment:
you'll actually need to copy the entire function approach here https://github.com/napari/napari/blob/master/napari/_vispy/vispy_image_layer.py#L144-L162 - it does two additional things, one is the `ensure_colormap_tuple` and the other is the `self._gamma` so we don't need the `layer.gamma` property anymore (which we will eventually want to remove).

You'll also need to do https://github.com/napari/napari/blob/master/napari/_vispy/vispy_image_layer.py#L178-L195 and add https://github.com/napari/napari/blob/master/napari/_vispy/vispy_image_layer.py#L25-L28.

I guess we could have an additional baseclass `VispyImageBaseLayer` that both inherit from if you want to reduce code duplication.

Reply:
Can you also add a doc string

Reply:
this is what confused me, the gamma is inherited from intensity_mixin.py, which seems different from what other layers do, is it a legacy component not updated or is there anything special for surface layer?

Reply:
I am updating the PR so I can move on, but I do want to know some history behind surface layere :)

Reply:
Yeah, so the properties in the `intensity_mixin` are shared between Image and Surface, and so we used an inheritance pattern to reduce code duplication. Looking forward I think we'll want to refactor to use a component pattern instead, which will make this much much nicer, for example both Image and Surface could then use a `ColorManger` or `RenderingManager` component that would control properties now in the `intensity_mixin`. Does that make more sense?